nomt/src/beatree/ops/update/branch_updater.rs

```diff
@@ -239,52 +375,72 @@ impl BranchUpdater {
 
         let mut left_gauge = BranchGauge::new();
         while left_gauge.body_size() < midpoint {
-            let (key, separator_len) = match self.ops[split_point] {
-                BranchOp::Keep(i, separator_len) => {
-                    // UNWRAP: keep ops require base to exist.
-                    let k = self.base.as_ref().unwrap().key(i);
-                    (k, separator_len)
+            match *&self.ops[split_point] {
+                BranchOp::Insert(key, _) => {
+                    if left_gauge.body_size_after(key, separator_len(&key)) > BRANCH_NODE_BODY_SIZE
+                    {
+                        if left_gauge.body_size() < BRANCH_MERGE_THRESHOLD {
+                            // rare case: body was artifically small due to long shared prefix.
+                            // start applying items without prefix compression. we assume items are less
+                            // than half the body size, so the next item should apply cleanly.
+                            left_gauge.stop_prefix_compression();
+                        } else {
+                            break;
+                        }
+                    }
                 }
-                BranchOp::Insert(k, _) => (k, separator_len(&k)),
-            };
-
-            if left_gauge.body_size_after(key, separator_len) > BRANCH_NODE_BODY_SIZE {
-                if left_gauge.body_size() < BRANCH_MERGE_THRESHOLD {
-                    // rare case: body was artifically small due to long shared prefix.
-                    // start applying items without prefix compression. we assume items are less
-                    // than half the body size, so the next item should apply cleanly.
-                    left_gauge.stop_prefix_compression();
-                } else {
-                    break;
+                BranchOp::Update(pos, _) => {
+                    let key = self.base.as_ref().unwrap().key(pos);
+                    if left_gauge.body_size_after(key, separator_len(&key)) > BRANCH_NODE_BODY_SIZE
+                    {
+                        if left_gauge.body_size() < BRANCH_MERGE_THRESHOLD {
+                            // same case as above
+                            left_gauge.stop_prefix_compression();
+                        } else {
+                            break;
+                        }
+                    }
                 }
-            }
+                BranchOp::KeepChunk(from, to) => {
+                    // try to split the chunk to make it fit into the available space
+                    let n_items = split_keep_chunk(
+                        self.base.as_ref().unwrap(),
+                        &left_gauge,
+                        &mut self.ops,
+                        split_point,
+                        midpoint,
+                        BRANCH_NODE_BODY_SIZE,
+                    );
+
+                    if n_items == 0 {
```

Comment:
I'm not clear why this is necessary; isn't `split_keep_chunk` meant to encapsulate this?

Reply:
No, this could become clearer if the function were renamed `try_split_keep_chunk`, as it accepts both a target and a limit that cannot be exceeded. The function's sole purpose is to split a chunk, which could potentially fail. For example, if the first item in the chunk requires prefix compression of the new node to stop.

This logic is also present in `bulk_split_step`: attempt to split a chunk; if it fails, change the first element of `KeepChunk` to an `Insert` and repeat the loop, potentially leading to the scenario where `left_gauge.stop_prefix_compression();` is reached.